service/topology_coordinator.cc

```diff
@@ -740,6 +755,88 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                    if (current_timestamp - generation > timeout && !_topo_sm._topology.contains(raft::server_id{host_id.id})) {
+                        topology_mutation_builder builder(guard.write_timestamp());
+                        // This topology mutation moves a node to left state and bans it. Hence, the value of below fields are not useful.
+                        // The dummy_value used for few fields indicates the trivialness of this row entry, and is used to detect this special case.
+                        static constexpr uint32_t dummy_value = 0;
+                        builder.with_node(raft::server_id{host_id.id})
+                                .set("datacenter", get_application_state_gently(eps.get_application_state_map(), gms::application_state::DC))
+                                .set("rack", get_application_state_gently(eps.get_application_state_map(), gms::application_state::RACK))
+                                .set("release_version", get_application_state_gently(eps.get_application_state_map(), gms::application_state::RELEASE_VERSION))
+                                .set("num_tokens", dummy_value)
+                                .set("tokens_string", sstring{})
+                                .set("shard_count", dummy_value)
+                                .set("ignore_msb", dummy_value)
+                                .set("request_id", dummy_value)
+                                .set("cleanup_status", cleanup_status::clean)
+                                .set("node_state", node_state::left);
+                        reason.append(::format(" {}/{},", host_id, addr));
+                        updates.push_back({builder.build()});
+                    }
+                });
+                if (updates.size()) {
+                    co_await update_topology_state(std::move(guard), std::move(updates), reason);
+                }
+
+            } catch (raft::request_aborted&) {
+                rtlogger.debug("gossiper orphan remover fiber aborted");
+            } catch (seastar::abort_requested_exception) {
+                rtlogger.debug("gossiper orphan remover fiber aborted");
+            } catch (group0_concurrent_modification&) {
+            } catch (term_changed_error&) {
+                rtlogger.debug("gossiper orphan remover fiber notices term change {} -> {}", _term, _raft.get_current_term());
+            } catch (...) {
+                rtlogger.error("gossiper orphan remover fiber got error {}", std::current_exception());
+            }
+            if (do_speedup_fiber) {
+                try {
+                    co_await seastar::sleep_abortable(std::chrono::milliseconds(1), _as);
+                } catch (...) {
+                    rtlogger.debug("gossiper orphan remover: sleep failed: {}", std::current_exception());
+                }
+                continue;
+            }
+            try {
+                co_await seastar::sleep_abortable(std::chrono::seconds(10), _as);
```

Comment:
Cannot we wait in some event rather than waking up periodically?

Reply:
It seems like event based mechanisms can increase complexity. For example, one event based solution I can think of is that during joining, if collision is detected, then we can perform checks with the collided ip, but this solution involves sending remove request to leader and timeout checks on leader, making the solution complex.

Secondly, the cpu utilized by this thread is anyway very less(per 10 secs), and this method seems simple ~100 lines of code. So I am inclined towards this.

Thoughts?

Reply:
I can't say I like it but don't know enough to suggest a good event.